Add Chat UI functionality

shacharlevy opened the pull request:
This PR contains:
- New chat UI with courses and study groups messages display and sending.
- Rest API for getting and posting messages (using djangorestframework).
- Tests for the API. 
 
![image](https://github.com/redhat-beyond/HighTeach/assets/51419772/df478498-1245-4a62-bc4c-a7168b7916e7)

fixes #102

shirazyomtov commented on chat/templates/chat/chat.html:
```diff
@@ -0,0 +1,68 @@
+    var staticPhotosFolder = "{% static "images/" %}";
+    var currentUserID = {{request.user.id}};
+    var csrf_token = '{{ csrf_token }}';
+</script>
```
Maybe it can be moved to one of the JS files

shacharlevy replied:
As I found this is the option that is available in order to use django variables in a separated JS file (the assignment must be in the template for it to render)

shirazyomtov commented on static/styles/chat.css:
```diff
@@ -0,0 +1,283 @@
+
+.clickable {
+  cursor: pointer;
+}
```
Missing new line in the end of the file

shacharlevy replied:
Done

shirazyomtov reviewed (changes requested):
Great work, I added two comments

shacharlevy commented on HighTeach/urls.py:
```diff
@@ -26,4 +27,8 @@
     path('logout/', auth_views.LogoutView.as_view(template_name='main/homePage.html'), name='logout'),
     path('study-group/', include('study_group.urls')),
     path('course/', include('course.urls')),
+    path('chat', chat_views.chat_view, name='chat'),
```
Done

shacharlevy commented on chat/models.py:
```diff
@@ -12,6 +12,17 @@ def get_group_chat(self, group_id):
 
+    @staticmethod
+    def get_student_contacts_list(user_id):
+        student_courses = StudentCourse.objects.filter(student_id=user_id)
```
Done

shacharlevy commented on chat/models.py:
```diff
@@ -12,6 +12,17 @@ def get_group_chat(self, group_id):
+        return student_courses
+
+    @staticmethod
+    def get_teacher_contacts_list(user_id):
```
Done

shacharlevy commented on chat/views.py:
```diff
@@ -1,3 +1,51 @@
 # Create your views here.
+def chat_view(request):
+    if request.user.profile.account_type == 'S':
+        student_courses = ChatManager.get_student_contacts_list(request.user)
```
Done

shirazyomtov reviewed (approved):
LGTM

natifridman reviewed (commented):
Out of curiosity, why did you use Django rest instead of something like in the [msgboard](https://github.com/redhat-beyond/beyond-tutorial/blob/main/msgboard/views.py)?

Also, could we add some tests here? what about tests for unauthorized or guest users? or checking the response status code?

shacharlevy commented:
> Out of curiosity, why did you use Django rest instead of something like in the [msgboard](https://github.com/redhat-beyond/beyond-tutorial/blob/main/msgboard/views.py)?
> 
> Also, could we add some tests here? what about tests for unauthorized or guest users? or checking the response status code?

I added and updated the tests.
The reason for Django rest is because the page is dynamic, I needed to know which chat is currently selected so I used js and then since I had all the data in the js I wanted to create a post request (to add a new message) and this is how I found out about django rest.

natifridman commented on tests/chat/test_chat.py:
```diff
@@ -61,3 +61,111 @@ def test_get_student_course_chat(self, message_with_student_course):
+        chat_messages = Message.objects.get_student_course_chat(student_course0.student_course_id)
+        len_messages_before_post = len(chat_messages)
+        client.force_login(user=persist_second_user)
+        client.post('/chat/post-message/', {
+            "courseId": student_course0.student_course_id,
+            "message": "Test",
+        })
```
Maybe the message can be a fixture so we won't need to repeat this?

shacharlevy replied:
Done

natifridman reviewed (commented):
Nice work, added one comment

Omeramsc commented on chat/views.py:
```diff
@@ -1,3 +1,64 @@
+from rest_framework.decorators import api_view
+
 
 # Create your views here.
```
remove comment

shacharlevy replied:
Done

Omeramsc commented on chat/views.py:
```diff
@@ -1,3 +1,64 @@
+    if 'courseId' in request.data:
+        course = StudentCourse.objects.get_course(int(request.data['courseId']))[0]
+    else:
+        course = None
```
You can also declare the course variable before the conditional statement, removing the need for the `else`
Also, is trying to get `StudentCourse.objects.get_course(int(request.data['courseId']))` fails when there is no `courseId` in request.data? or does it just return None? if so, you can remove the whole if (or add a `or none` at the end of the assignment).

shacharlevy replied:
Changed (removed the else clause). if 'courseId' is not in request.data, the above call will raise an exception (when calling request.data['courseId']), so the if is still needed. Thanks to this comment I noticed that get_course can return an empty list (if it contains bad courseId), so I modified it to check for this case.

Omeramsc commented on chat/views.py:
```diff
@@ -1,3 +1,64 @@
+    else:
+        course = None
+    if 'groupId' in request.data:
+        group = StudyGroup.objects.get_group_by_id(int(request.data['groupId']))[0]
```
same

shacharlevy replied:
Done

Omeramsc commented on chat/views.py:
```diff
@@ -1,3 +1,64 @@
+    try:
+        new_message = Message(sender=request.user, student_course=course, group=group, message=request.data['message'])
+        new_message.save()
+    except ValidationError:
```
what would make it raise?

shacharlevy replied:
The new_message.save() will raise ValidationError if for example both the group and the course are None (due to the Message constraints I implemented in an old PR).

Omeramsc replied:
If so I'd make it more visible in the test. either by variable names or the test's name  "test_x_raise_when_y_z"

shacharlevy replied:
The tests in this PR are for the API endpoint (sending an http request with the bad arguments) and therefore cannot assert for a raised exception in the server side, just the obtained HTTP response and database state (to make sure the operation was not performed for bad input). There are other tests from an older PR that check for a ValidationError when calling meesage.save() for an invalid message (in the TestMessage class of test_chat.py).

Omeramsc replied:
OK, but I don't understand how that's relevant to naming conventions and readability?
Maybe there was a miscommunication on my side; I meant that the way the test was written got me wondering what is it about and why. the answer you provided me is fine, but it only resides in Github. in order to improve it for future engineers dealing with the code, I'd suggest renaming the test in a way that will make it clear: `<what is being tested>-<expected-result>-<why\when-is-it-expected>`

also, you can rename `new_message` to `new_message_with_no_group_or_course`

shacharlevy replied:
I think there is a misunderstanding, the new_message is a variable in the tested method - it can be both valid and invalid depending on the sent data. The tests themselves are currently named somewhat like the format you suggested, for example `test_add_message_to_group_chat_not_logged_in` tests adding message to a group chat, without a logged in user, and the test defines the expected result which is a response with status code 401 and that no message was added.

Omeramsc commented on tests/chat/test_chat.py:
```diff
@@ -61,3 +61,111 @@ def test_get_student_course_chat(self, message_with_student_course):
+        assert len(chat_messages_after_post) == len_messages_before_post
+        assert response.status_code == 401
+
+    def test_add_message_to_group_chat_not_in_group(self, client, persist_second_user, study_group0):
```
is this scenario even possible? as `test_get_study_group_chat_not_in_group` shows us you cannot get to the group chat if you are not a part of the group

shacharlevy replied:
The UI does not enable such scenario. This test checks for a scenario where a user tries to access the API endpoint not through the UI (and this is a different endpoint from the get chat endpoint).

Omeramsc commented on tests/chat/test_chat.py:
```diff
@@ -61,3 +61,111 @@ def test_get_student_course_chat(self, message_with_student_course):
+        chat_messages_after_post = Message.objects.get_group_chat(study_group0.study_group_id)
+        assert len(chat_messages_after_post) == len_messages_before_post + 1
+
+    def test_add_message_to_group_chat_not_logged_in(self, client, study_group0):
```
same

shacharlevy replied:
The UI does not enable such scenario. This test checks for a scenario where a user tries to access the API endpoint not through the UI (and this is a different endpoint from the get chat endpoint).

natifridman reviewed (approved):
LGTM

shirazyomtov reviewed (approved):
LGTM

Omeramsc reviewed (approved):
LGTM
